File: dokka/private/DokkaGeneratorRunner.java
package rules.dokka;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import org.jetbrains.dokka.ConfigurationJsonUtilsKt;
import org.jetbrains.dokka.ConfigurationKt;
import org.jetbrains.dokka.DokkaConfiguration;
import org.jetbrains.dokka.DokkaConfigurationImpl;
import org.jetbrains.dokka.DokkaGenerator;
import org.jetbrains.dokka.DokkaModuleDescriptionImpl;
import org.jetbrains.dokka.LinkMapperKt;
import org.jetbrains.dokka.PluginConfigurationImpl;
import org.jetbrains.dokka.utilities.DokkaConsoleLogger;

/**
 * Runs Dokka in-process with paths resolved inside the Bazel action's execution root.
 *
 * <p>Normal configurations use Dokka's documented JSON parser. Multi-module manifests are
 * converted to typed module descriptions because Dokka does not document the aggregate {@code
 * modules} JSON representation.
 */
public final class DokkaGeneratorRunner {
  private static final String MULTI_MODULE_MODE = "rules_dokka_multi_module";
  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

  public static void main(String[] args) throws Exception {
    if (args.length != 1) {
      throw new IllegalArgumentException("usage: DokkaGeneratorRunner CONFIGURATION");
    }

    ObjectNode root = (ObjectNode) OBJECT_MAPPER.readTree(Paths.get(args[0]).toFile());
    DokkaConfiguration configuration =
        MULTI_MODULE_MODE.equals(root.path("runnerMode").asText())
            ? multiModuleConfiguration(root)
            : standardConfiguration(root);
    new DokkaGenerator(configuration, new DokkaConsoleLogger()).generate();
  }

  private static DokkaConfiguration standardConfiguration(ObjectNode root) throws Exception {
    absolutizeField(root, "outputDir");
    absolutizeOptionalField(root, "cacheRoot");
    absolutizeArray(root, "includes");
    absolutizeArray(root, "pluginsClasspath");

    JsonNode sourceSets = root.path("sourceSets");
    if (sourceSets.isArray()) {
      for (JsonNode sourceSetNode : (ArrayNode) sourceSets) {
        ObjectNode sourceSet = (ObjectNode) sourceSetNode;
        for (String field :
            List.of(
                "classpath",
                "includes",
                "samples",
                "sourceRoots",
                "suppressedFiles")) {
          absolutizeArray(sourceSet, field);
        }
        JsonNode sourceLinks = sourceSet.path("sourceLinks");
        if (sourceLinks.isArray()) {
          for (JsonNode sourceLinkNode : (ArrayNode) sourceLinks) {
            absolutizeOptionalField((ObjectNode) sourceLinkNode, "localDirectory");
          }
        }
      }
    }

    String json = OBJECT_MAPPER.writeValueAsString(root);
    DokkaConfiguration configuration =
        ConfigurationKt.apply(
            ConfigurationJsonUtilsKt.DokkaConfigurationImpl(json),
            ConfigurationJsonUtilsKt.GlobalDokkaConfiguration(json));
    for (DokkaConfiguration.DokkaSourceSet sourceSet : configuration.getSourceSets()) {
      sourceSet.getExternalDocumentationLinks().addAll(LinkMapperKt.defaultLinks(sourceSet));
    }
    return configuration;
  }

  private static DokkaConfiguration multiModuleConfiguration(ObjectNode root) {
    Path outputDirectory = absolutePath(requiredText(root, "outputDir"));
    List<DokkaModuleDescriptionImpl> modules = new ArrayList<>();
    JsonNode moduleNodes = root.path("modules");
    if (!moduleNodes.isArray() || moduleNodes.isEmpty()) {
      throw new IllegalArgumentException("Multi-module Dokka manifest must contain modules");
    }

    for (JsonNode moduleNode : (ArrayNode) moduleNodes) {
      ObjectNode module = (ObjectNode) moduleNode;
      Path relativeOutput = Paths.get(requiredText(module, "relativePathToOutputDirectory"));
      if (relativeOutput.isAbsolute()
          || !outputDirectory.resolve(relativeOutput).normalize().startsWith(outputDirectory)) {
        throw new IllegalArgumentException(
            "Module output escapes the aggregate documentation directory: " + relativeOutput);
      }
      modules.add(
          new DokkaModuleDescriptionImpl(
              requiredText(module, "name"),
              relativeOutput.toFile(),
              files(module.path("includes")),
              absolutePath(requiredText(module, "sourceOutputDirectory")).toFile()));
    }

    List<PluginConfigurationImpl> pluginConfigurations = new ArrayList<>();
    JsonNode pluginNodes = root.path("pluginsConfiguration");
    if (pluginNodes.isArray()) {
      for (JsonNode pluginNode : (ArrayNode) pluginNodes) {
        pluginConfigurations.add(
            new PluginConfigurationImpl(
                requiredText(pluginNode, "fqPluginName"),
                DokkaConfiguration.SerializationFormat.valueOf(
                    requiredText(pluginNode, "serializationFormat")),
                requiredText(pluginNode, "values")));
      }
    }

    return new DokkaConfigurationImpl(
        requiredText(root, "moduleName"),
        optionalText(root, "moduleVersion"),
        outputDirectory.toFile(),
        optionalFile(root, "cacheRoot"),
        root.path("offlineMode").asBoolean(),
        List.of(),
        new ArrayList<>(files(root.path("pluginsClasspath"))),
        pluginConfigurations,
        modules,
        root.path("failOnWarning").asBoolean(),
        false,
        root.path("suppressObviousFunctions").asBoolean(true),
        files(root.path("includes")),
        root.path("suppressInheritedMembers").asBoolean(),
        root.path("finalizeCoroutines").asBoolean(true));
  }

  private static Set<File> files(JsonNode nodes) {
    Set<File> files = new LinkedHashSet<>();
    if (nodes.isArray()) {
      for (JsonNode node : (ArrayNode) nodes) {
        files.add(absolutePath(node.asText()).toFile());
      }
    }
    return files;
  }

  private static void absolutizeArray(ObjectNode node, String field) {
    JsonNode values = node.path(field);
    if (!values.isArray()) {
      return;
    }
    ArrayNode absoluteValues = OBJECT_MAPPER.createArrayNode();
    for (JsonNode value : (ArrayNode) values) {
      absoluteValues.add(absolutePath(value.asText()).toString());
    }
    node.set(field, absoluteValues);
  }

  private static void absolutizeField(ObjectNode node, String field) {
    node.put(field, absolutePath(requiredText(node, field)).toString());
  }

  private static void absolutizeOptionalField(ObjectNode node, String field) {
    String value = optionalText(node, field);
    if (value != null && !value.isEmpty()) {
      node.put(field, absolutePath(value).toString());
    }
  }

  private static File optionalFile(ObjectNode node, String field) {
    String value = optionalText(node, field);
    return value == null || value.isEmpty() ? null : absolutePath(value).toFile();
  }

  private static String requiredText(JsonNode node, String field) {
    String value = optionalText(node, field);
    if (value == null || value.isEmpty()) {
      throw new IllegalArgumentException("Missing required Dokka manifest field: " + field);
    }
    return value;
  }

  private static String optionalText(JsonNode node, String field) {
    JsonNode value = node.get(field);
    return value == null || value.isNull() ? null : value.asText();
  }

  private static Path absolutePath(String path) {
    return Paths.get(path).toAbsolutePath().normalize();
  }

  private DokkaGeneratorRunner() {}
}

File: dokka/private/DokkaGeneratorRunner.kt
package rules.dokka

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode
import org.jetbrains.dokka.DokkaConfiguration
import org.jetbrains.dokka.DokkaConfigurationImpl
import org.jetbrains.dokka.DokkaGenerator
import org.jetbrains.dokka.DokkaModuleDescriptionImpl
import org.jetbrains.dokka.GlobalDokkaConfiguration
import org.jetbrains.dokka.PluginConfigurationImpl
import org.jetbrains.dokka.apply
import org.jetbrains.dokka.defaultLinks
import org.jetbrains.dokka.utilities.DokkaConsoleLogger
import java.io.File
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Runs Dokka in-process with paths resolved inside the Bazel action's execution root.
 *
 * Normal configurations use Dokka's documented JSON parser. Multi-module manifests are converted
 * to typed module descriptions because Dokka does not document the aggregate `modules` JSON
 * representation.
 */
object DokkaGeneratorRunner {
    private const val MULTI_MODULE_MODE = "rules_dokka_multi_module"
    private val objectMapper = ObjectMapper()

    @JvmStatic
    fun main(args: Array<String>) {
        require(args.size == 1) { "usage: DokkaGeneratorRunner CONFIGURATION" }

        val root = objectMapper.readTree(Paths.get(args[0]).toFile()) as ObjectNode
        val configuration =
            if (root.path("runnerMode").asText() == MULTI_MODULE_MODE) {
                multiModuleConfiguration(root)
            } else {
                standardConfiguration(root)
            }
        DokkaGenerator(configuration, DokkaConsoleLogger()).generate()
    }

    private fun standardConfiguration(root: ObjectNode): DokkaConfiguration {
        absolutizeField(root, "outputDir")
        absolutizeOptionalField(root, "cacheRoot")
        absolutizeArray(root, "includes")
        absolutizeArray(root, "pluginsClasspath")

        val sourceSets = root.path("sourceSets")
        if (sourceSets.isArray) {
            sourceSets.forEach { sourceSetNode ->
                val sourceSet = sourceSetNode as ObjectNode
                listOf(
                    "classpath",
                    "includes",
                    "samples",
                    "sourceRoots",
                    "suppressedFiles",
                ).forEach { field -> absolutizeArray(sourceSet, field) }

                val sourceLinks = sourceSet.path("sourceLinks")
                if (sourceLinks.isArray) {
                    sourceLinks.forEach { sourceLinkNode ->
                        absolutizeOptionalField(sourceLinkNode as ObjectNode, "localDirectory")
                    }
                }
            }
        }

        val json = objectMapper.writeValueAsString(root)
        val configuration =
            DokkaConfigurationImpl(json).apply(GlobalDokkaConfiguration(json))
        configuration.sourceSets.forEach { sourceSet ->
            @Suppress("UNCHECKED_CAST")
            val externalDocumentationLinks =
                sourceSet.externalDocumentationLinks
                    as MutableSet<DokkaConfiguration.ExternalDocumentationLink>
            externalDocumentationLinks.addAll(defaultLinks(sourceSet))
        }
        return configuration
    }

    private fun multiModuleConfiguration(root: ObjectNode): DokkaConfiguration {
        val outputDirectory = absolutePath(requiredText(root, "outputDir"))
        val moduleNodes = root.path("modules")
        require(moduleNodes.isArray && !moduleNodes.isEmpty) {
            "Multi-module Dokka manifest must contain modules"
        }

        val modules =
            moduleNodes.map { moduleNode ->
                val module = moduleNode as ObjectNode
                val portableRelativeOutput =
                    requiredText(module, "relativePathToOutputDirectory")
                val relativeOutput = Paths.get(portableRelativeOutput)
                require(
                    !relativeOutput.isAbsolute &&
                        outputDirectory.resolve(relativeOutput).normalize().startsWith(outputDirectory),
                ) {
                    "Module output escapes the aggregate documentation directory: $relativeOutput"
                }
                DokkaModuleDescriptionImpl(
                    requiredText(module, "name"),
                    PortableModulePath(relativeOutput, portableRelativeOutput),
                    files(module.path("includes")),
                    absolutePath(requiredText(module, "sourceOutputDirectory")).toFile(),
                )
            }

        val pluginNodes = root.path("pluginsConfiguration")
        val pluginConfigurations =
            if (pluginNodes.isArray) {
                pluginNodes.map { pluginNode ->
                    PluginConfigurationImpl(
                        requiredText(pluginNode, "fqPluginName"),
                        DokkaConfiguration.SerializationFormat.valueOf(
                            requiredText(pluginNode, "serializationFormat"),
                        ),
                        requiredText(pluginNode, "values"),
                    )
                }
            } else {
                emptyList()
            }

        // AllModulesPageGeneration ignores the single-module warning and suppression settings.
        return DokkaConfigurationImpl(
            requiredText(root, "moduleName"),
            optionalText(root, "moduleVersion"),
            outputDirectory.toFile(),
            optionalFile(root, "cacheRoot"),
            root.path("offlineMode").asBoolean(),
            emptyList(),
            files(root.path("pluginsClasspath")).toList(),
            pluginConfigurations,
            modules,
            false,
            false,
            true,
            files(root.path("includes")),
            false,
            root.path("finalizeCoroutines").asBoolean(true),
        )
    }

    private fun files(nodes: JsonNode): Set<File> =
        if (nodes.isArray) {
            nodes.mapTo(linkedSetOf()) { node -> absolutePath(node.asText()).toFile() }
        } else {
            emptySet()
        }

    private fun absolutizeArray(
        node: ObjectNode,
        field: String,
    ) {
        val values = node.path(field)
        if (!values.isArray) {
            return
        }
        val absoluteValues = objectMapper.createArrayNode()
        values.forEach { value -> absoluteValues.add(absolutePath(value.asText()).toString()) }
        node.set<ArrayNode>(field, absoluteValues)
    }

    private fun absolutizeField(
        node: ObjectNode,
        field: String,
    ) {
        node.put(field, absolutePath(requiredText(node, field)).toString())
    }

    private fun absolutizeOptionalField(
        node: ObjectNode,
        field: String,
    ) {
        optionalText(node, field)
            ?.takeIf(String::isNotEmpty)
            ?.let { value -> node.put(field, absolutePath(value).toString()) }
    }

    private fun optionalFile(
        node: ObjectNode,
        field: String,
    ): File? =
        optionalText(node, field)
            ?.takeIf(String::isNotEmpty)
            ?.let { value -> absolutePath(value).toFile() }

    private fun requiredText(
        node: JsonNode,
        field: String,
    ): String =
        requireNotNull(optionalText(node, field)?.takeIf(String::isNotEmpty)) {
            "Missing required Dokka manifest field: $field"
        }

    private fun optionalText(
        node: JsonNode,
        field: String,
    ): String? = node.get(field)?.takeUnless(JsonNode::isNull)?.asText()

    private fun absolutePath(path: String): Path = Paths.get(path).toAbsolutePath().normalize()

    /**
     * Preserves a URL-compatible module path when Dokka renders aggregate links.
     *
     * Dokka 2.2's all-modules-page plugin appends `relativePathToOutputDirectory` directly to a
     * string when it creates module URLs. A regular [File] renders that path with backslashes on
     * Windows. File operations still use the native path stored by [File]; only its string
     * representation remains portable.
     */
    private class PortableModulePath(
        nativePath: Path,
        private val portablePath: String,
    ) : File(nativePath.toString()) {
        override fun toString(): String = portablePath

        private companion object {
            private const val serialVersionUID = 1L
        }
    }
}
